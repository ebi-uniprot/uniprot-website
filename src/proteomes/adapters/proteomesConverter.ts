import { Lineage, Xref } from '../../shared/types/apiModel';
import { TaxonomyDatum } from '../../supporting-data/taxonomy/adapters/taxonomyConverter';
import { Citation } from '../../supporting-data/citations/adapters/citationsConverter';

export type GenomeAnnotation = {
  source: string;
  url?: string;
};

export type Component = {
  name: string;
  description: string;
  genomeAnnotation: GenomeAnnotation;
  proteomeCrossReferences?: Xref[];
  proteinCount: number; // used in the entry for each component TODO: eventually will be supported by backend in 2021_02 - 2021_03
};

export type CpdStatus =
  | 'Close to Standard'
  | 'Standard'
  | 'Outlier'
  | 'Unknown';

export type CpdReport = {
  proteomeCount: number;
  stdCdss: number;
  averageCdss: number;
  confidence: number;
  status: CpdStatus;
};

export type BuscoReport = {
  complete: number;
  completeSingle: number;
  completeDuplicated: number;
  fragmented: number;
  missing: number;
  total: number;
  lineageDb: string;
  score: number;
};

export type ProteomeCompletenessReport = {
  cpdReport?: CpdReport;
  buscoReport?: BuscoReport;
};

export type GenomeAssembly = {
  source: string;
  assemblyId: string;
  genomeAssemblyUrl?: string;
  level: string; // Genome representation
};

export type RedundantProteome = {
  id: string;
  similarity: number;
};

export type ProteomeType =
  | 'Reference and representative proteome'
  | 'Reference proteome'
  | 'Representative proteome'
  | 'Redundant proteome'
  | 'Other proteome'
  | 'Excluded';

export type ProteomesAPIModel = {
  id: string;
  taxonomy: TaxonomyDatum;
  modified: string;
  proteomeType: ProteomeType;
  components?: Component[];
  citations: Citation[];
  annotationScore: number;
  superkingdom: string;
  proteomeCompletenessReport: ProteomeCompletenessReport;
  genomeAssembly?: GenomeAssembly;
  geneCount: number;
  genomeAnnotation: GenomeAnnotation;
  taxonLineage: Lineage[];
  strain?: string;
  isolate?: string;
  panproteome?: string;
  description: string;
  redundantProteomes?: RedundantProteome[];
  redundantTo?: string;
  proteinCount: number; // use this in the results table - calculated sum of the components proteinCount: components.reduce((total, { proteinCount }) => proteinCount + total, 0)  TODO: eventually will be supported by backend in in 2021_02 - 2021_03
};

export type ProteomesUIModel = ProteomesAPIModel & {
  // any addition/change by the converter
};

const proteomesConverter = (data: ProteomesAPIModel): ProteomesUIModel => ({
  ...data,
});

export default proteomesConverter;
