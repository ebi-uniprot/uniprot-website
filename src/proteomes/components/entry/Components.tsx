import { Fragment, FC, ReactNode } from 'react';
import { Link } from 'react-router-dom';
import { Card, DataTable, ExternalLink } from 'franklin-sites';

import useItemSelect from '../../../shared/hooks/useItemSelect';

import ComponentsButtons from './ComponentsButtons';

import externalUrls from '../../../shared/config/externalUrls';
import { LocationToPath, Location } from '../../../app/config/urls';

import {
  ProteomesAPIModel,
  Component,
} from '../../adapters/proteomesConverter';

import '../styles/overview.scss';

const genomeAccessionDB = 'GenomeAccession' as const;

export const Components: FC<Pick<ProteomesAPIModel, 'components' | 'id'>> = ({
  components,
  id,
}) => {
  const [selectedEntries, handleItemSelection] = useItemSelect();

  if (!components?.length) {
    return null;
  }

  const columns: Array<{
    label: ReactNode;
    name: string;
    render: (arg: Component) => ReactNode;
    width?: string;
    ellipsis?: boolean;
  }> = [
    {
      label: 'Component name',
      name: 'component_name',
      render: ({ name }) => name,
    },
    {
      label: 'Genome accession(s)',
      name: 'genome_accession',
      render: ({ proteomeCrossReferences }) =>
        proteomeCrossReferences
          ?.filter(({ database, id }) => id && database === genomeAccessionDB)
          .map(({ id }, index) => (
            <Fragment key={id}>
              {index ? ', ' : ''}
              <ExternalLink
                url={externalUrls.ENABrowser(id as string)}
                key={id}
              >
                {id}
              </ExternalLink>
            </Fragment>
          )),
    },
    {
      label: 'Proteins',
      name: 'proteins',
      render: ({ proteinCount, name }) => (
        <Link
          to={{
            pathname: LocationToPath[Location.UniProtKBResults],
            search: `query=(proteome:${id}) AND (proteomecomponent:"${name}")`,
          }}
        >
          {proteinCount || 'no data yet'}
        </Link>
      ),
    },
  ];

  return (
    <Card title="Components">
      <ComponentsButtons
        nComponents={components.length}
        selectedEntries={selectedEntries}
        id={id}
      />
      <DataTable
        getIdKey={({ name }) => name}
        density="compact"
        columns={columns}
        data={components}
        selected={selectedEntries}
        onSelectRow={handleItemSelection}
        fixedLayout
      />
    </Card>
  );
};
export default Components;
