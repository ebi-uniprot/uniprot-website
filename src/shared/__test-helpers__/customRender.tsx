/**
 * This tries to follow what is detailed here:
 * https://testing-library.com/docs/react-testing-library/setup#custom-render
 */
import { render, RenderOptions } from '@testing-library/react';
import { Component, Dispatch, ReactElement, ReactNode } from 'react';
import { HelmetProvider } from 'react-helmet-async';
import {
  createMemoryHistory,
  LocationState,
  MemoryHistory,
  Route,
  Router,
} from 'react-router';
import { JsonValue, SetRequired } from 'type-fest';

import { MessagesState } from '../../messages/state/messagesInitialState';
import { MessagesAction } from '../../messages/state/messagesReducers';
import databaseInfo from '../../uniprotkb/utils/__tests__/__mocks__/databaseInfo';
import { getDatabaseInfoMaps } from '../../uniprotkb/utils/database';
import {
  MessagesDispatchContext,
  MessagesStateContext,
} from '../contexts/Messages';
import {
  DatabaseInfoMapsContext,
  UniProtDataVersionContext,
} from '../contexts/UniProtData';
import { JobAction } from '../workers/jobs/state/actionHandler';
import { JobsState } from '../workers/jobs/state/jobsInitialState';

type ExtraRenderOptions = {
  // For react-router
  history?: MemoryHistory<LocationState>;
  path?: string;
  // For custom user preferences (used by useLocalStorage)
  initialLocalStorage?: Record<string, JsonValue>;
  // For app context
  messagesState?: MessagesState;
  messagesDispatch?: jest.Mock<Dispatch<MessagesAction>>;
  JobsState?: JobsState;
  toolsDispatch?: jest.Mock<Dispatch<JobAction>>;
};

type WrapperProps = { children: ReactNode } & RenderOptions &
  SetRequired<
    ExtraRenderOptions,
    | 'history'
    | 'initialLocalStorage'
    | 'messagesState'
    | 'messagesDispatch'
    | 'JobsState'
    | 'toolsDispatch'
  >;

const dbInfoMaps = getDatabaseInfoMaps(databaseInfo);
const uniProtDataVersion = {
  releaseNumber: '2023_01',
  releaseDate: new Date('01-March-2023'),
};

class Wrapper extends Component<WrapperProps> {
  constructor(props: WrapperProps) {
    super(props);
    for (const [key, value] of Object.entries(props.initialLocalStorage)) {
      window.localStorage.setItem(key, JSON.stringify(value));
    }
  }

  // eslint-disable-next-line class-methods-use-this
  componentWillUnmount() {
    window.localStorage.clear();
  }

  render() {
    const { children, path, history, messagesDispatch, messagesState } =
      this.props;
    return (
      <HelmetProvider>
        <Router history={history}>
          <MessagesDispatchContext.Provider value={messagesDispatch}>
            <MessagesStateContext.Provider value={messagesState}>
              <DatabaseInfoMapsContext.Provider value={dbInfoMaps}>
                <UniProtDataVersionContext.Provider value={uniProtDataVersion}>
                  {path ? (
                    <Route path={path} render={() => children} />
                  ) : (
                    children
                  )}
                </UniProtDataVersionContext.Provider>
              </DatabaseInfoMapsContext.Provider>
            </MessagesStateContext.Provider>
          </MessagesDispatchContext.Provider>
        </Router>
      </HelmetProvider>
    );
  }
}

const customRender = (
  ui: ReactElement,
  {
    route = '',
    path,
    history = createMemoryHistory({ initialEntries: [route] }),
    initialLocalStorage = {},
    messagesState = {},
    messagesDispatch = jest.fn(),
    JobsState = {},
    toolsDispatch = jest.fn(),
    ...options
  }: RenderOptions &
    ExtraRenderOptions & {
      route?: string;
    } = {}
) => ({
  ...render(ui, {
    wrapper: (props) => (
      <Wrapper
        path={path}
        history={history}
        initialLocalStorage={initialLocalStorage}
        messagesState={messagesState}
        messagesDispatch={messagesDispatch}
        JobsState={JobsState}
        toolsDispatch={toolsDispatch}
        {...props}
      />
    ),
    ...options,
  }),
  history,
  messagesDispatch,
  toolsDispatch,
});

export default customRender;
