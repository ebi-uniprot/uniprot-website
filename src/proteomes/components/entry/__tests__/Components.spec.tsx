import customRender from '../../../../shared/__test-helpers__/customRender';

import Components from '../Components';

import proteomesConverter from '../../../adapters/proteomesConverter';

import mockData from '../../../__mocks__/proteomesEntryModelData';

describe('Components view', () => {
  it('should render', () => {
    const data = proteomesConverter(mockData);
    const { asFragment } = customRender(
      <Components
        components={data.components}
        id={data.id}
        proteinCount={data.proteinCount}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
