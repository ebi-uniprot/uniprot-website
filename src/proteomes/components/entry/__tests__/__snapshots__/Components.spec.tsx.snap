// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components view should render 1`] = `
<DocumentFragment>
  <section
    class="card"
  >
    <div>
      <div
        class="card__header"
      >
        <h2
          class="card__title"
        >
          Components
        </h2>
      </div>
      <div
        class="card__content"
      >
        <div
          class="button-group"
        >
          <button
            class="button tertiary"
            type="button"
          >
            <test-file-stub />
            Download
          </button>
          <a
            class="button tertiary"
            href="/uniprotkb?query=(proteome:UP000005640)"
          >
            View UniProtKB entry for component
          </a>
        </div>
        <table
          class="data-table data-table--compact data-table--fixed"
        >
          <thead>
            <tr>
              <th
                class="data-table__header-cell--checkbox"
              >
                 
              </th>
              <th
                class=""
              >
                Component name
              </th>
              <th
                class=""
              >
                Genome accession(s)
              </th>
              <th
                class=""
              >
                Proteins
              </th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td>
                <input
                  aria-label="name"
                  data-id="name"
                  type="checkbox"
                />
              </td>
              <td
                class="data-table__cell--ellipsis"
              >
                name
              </td>
              <td
                class="data-table__cell--ellipsis"
              >
                <a
                  class="external-link"
                  href="//www.ebi.ac.uk/ena/browser/view/ACA121"
                  rel="noopener"
                  target="_blank"
                >
                  ACA121
                  <test-file-stub
                    data-testid="external-link-icon"
                    width="12.5"
                  />
                </a>
                , 
                <a
                  class="external-link"
                  href="//www.ebi.ac.uk/ena/browser/view/ACA122"
                  rel="noopener"
                  target="_blank"
                >
                  ACA122
                  <test-file-stub
                    data-testid="external-link-icon"
                    width="12.5"
                  />
                </a>
              </td>
              <td
                class="data-table__cell--ellipsis"
              >
                <a
                  href="/uniprotkb?query=(proteome:UP000005640) AND (proteomecomponent:\\"name\\")"
                >
                  18
                </a>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </section>
</DocumentFragment>
`;
